### PR TITLE
GH991: Added IsDependentOn overload which takes CakeTaskBuilder<T> parameter.

### LINKED ISSUE #991: IsDependentOn with CakeTaskBuilder parameter
`IsDependentOn` takes a task name as a parameter which is prone to error. A better approach is to add an overload of `IsDependentOn` which takes existing `CakeTaskBuilder` as a parameter.

[Discussion on Gitter](https://gitter.im/cake-build/cake?at=57601dec814a5680714314e4)

### PR DESCRIPTION
**Pull request by Giorgi**
Resolves #991

### CODE REVIEW

**Comment by Giorgi** [comment]
@gep13 I created a new PR from feature branch.

File: src/Cake.Core/CakeTaskBuilderExtensions.cs [hunk 1]
```diff
@@ -30,6 +30,26 @@ public static CakeTaskBuilder<T> IsDependentOn<T>(this CakeTaskBuilder<T> builde
...
         /// <summary>
+        /// Creates a dependency between two tasks.
+        /// </summary>
+        /// <typeparam name="T">The task type.</typeparam>
+        /// <typeparam name="TOther">The task type that this task depends on.</typeparam>
+        /// <param name="builder">The task builder.</param>
+        /// <param name="taskBuilder">The name of the dependent task.</param>
+        /// <returns>The same <see cref="CakeTaskBuilder{T}"/> instance so that multiple calls can be chained.</returns>
+        public static CakeTaskBuilder<T> IsDependentOn<T, TOther>(this CakeTaskBuilder<T> builder, CakeTaskBuilder<TOther> taskBuilder)
+            where T : CakeTask where TOther : CakeTask
```
Comment (patriksvensson) [inline src/Cake.Core/CakeTaskBuilderExtensions.cs]: Could you separate this into two lines?

```charp
where T : CakeTask
where TOther : CakeTask
```

File: src/Cake.Core/CakeTaskBuilderExtensions.cs [hunk 2]
```diff
@@ -30,6 +30,26 @@ public static CakeTaskBuilder<T> IsDependentOn<T>(this CakeTaskBuilder<T> builde
...
 
         /// <summary>
+        /// Creates a dependency between two tasks.
+        /// </summary>
+        /// <typeparam name="T">The task type.</typeparam>
+        /// <typeparam name="TOther">The task type that this task depends on.</typeparam>
+        /// <param name="builder">The task builder.</param>
+        /// <param name="taskBuilder">The name of the dependent task.</param>
+        /// <returns>The same <see cref="CakeTaskBuilder{T}"/> instance so that multiple calls can be chained.</returns>
+        public static CakeTaskBuilder<T> IsDependentOn<T, TOther>(this CakeTaskBuilder<T> builder, CakeTaskBuilder<TOther> taskBuilder)
```
Comment (patriksvensson) [inline src/Cake.Core/CakeTaskBuilderExtensions.cs]: Could you rename `taskBuilder` to `other` so it's consistent?

File: src/Cake.Core/CakeTaskBuilderExtensions.cs [hunk 3]
```diff
@@ -30,6 +30,26 @@ public static CakeTaskBuilder<T> IsDependentOn<T>(this CakeTaskBuilder<T> builde
...
+            where T : CakeTask where TOther : CakeTask
+        {
+            if (builder == null)
```
Comment (patriksvensson) [inline src/Cake.Core/CakeTaskBuilderExtensions.cs]: Could you add a test that verifies that the method will throw if `builder` is null.

File: src/Cake.Core/CakeTaskBuilderExtensions.cs [hunk 4]
```diff
@@ -30,6 +30,26 @@ public static CakeTaskBuilder<T> IsDependentOn<T>(this CakeTaskBuilder<T> builde
...
+            {
+                throw new ArgumentNullException("builder");
+            }
+
```
Comment (patriksvensson) [inline src/Cake.Core/CakeTaskBuilderExtensions.cs]: Should check that `taskBuilder` isn't null as well.

**Comment by patriksvensson** [comment]
@Giorgi I've left some comments for you. Other than that I think it looks good :+1:

**Comment by Giorgi** [comment]
@patriksvensson Done.

**Comment by Giorgi** [comment]
@patriksvensson Anything else I can do?

**Comment by patriksvensson** [comment]
@Giorgi Sorry for not getting back to you sooner. I'm currently on vacation and my internet connection is very limited but will try to take a look at this as soon as I'm at a computer. Sorry for the inconvenience!

Could you rebase this PR against develop branch and squash the commits to a single one, then either me, @devlead or @gep13 will take it from there.

Sorry again...

**Comment by devlead** [comment]
@Giorgi would you be in an position to squash commits and rebase against latest develop?

**Comment by Giorgi** [comment]
@devlead Probably not soon. I think I messed up the repository trying to squash it last time.

**Comment by devlead** [comment]
@Giorgi ok, no worries, ping us when you have something you want us to review.

**Comment by gep13** [comment]
@Giorgi if you need some help getting the rebase/squash done, feel free to reach out as well, and we can hopefully help out.

**Comment by gep13** [comment]
@Giorgi thank you very much for your contribution to Cake, and apologies for this taking so long to be included.

I have now rebased and merged your changes here: https://github.com/cake-build/cake/commit/2fcf2f5f5383ee3c345962d95b3b3307e5d1a048

Due to the fact that a manual rebase was required, I am going to have to manually close this PR, however, correct attribution has been maintained.  Thanks again!

**Comment by Giorgi** [comment]
@gep13 Thanks a lot, it's very nice from you.

**Comment by gep13** [comment]
@Giorgi would welcome any help that you can offer here: https://github.com/cake-build/website/issues/193

Perhaps to update this page: http://cakebuild.net/docs/fundamentals/dependencies

Thanks!
